fs/backgroundfetcher/resolver.go

```diff
@@ -0,0 +1,102 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package backgroundfetcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	sm "github.com/awslabs/soci-snapshotter/fs/span-manager"
+	"github.com/awslabs/soci-snapshotter/soci"
+	"github.com/containerd/containerd/log"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+)
+
+type Resolver interface {
+	// Resolve fetches and caches the next span. Returns true if there is still more data to be fetched.
+	// Returns false otherwise.
+	Resolve(context.Context) (bool, error)
+
+	// Closes the resolver.
+	Close() error
+
+	// Checks whether the resolver is closed or not.
+	Closed() bool
+}
+
+type base struct {
+	*sm.SpanManager
+	layerDigest digest.Digest
+	closed      bool
+	closedMu    sync.Mutex
+}
+
+func (b *base) Close() error {
+	b.closedMu.Lock()
+	defer b.closedMu.Unlock()
+	b.closed = true
+	return nil
+}
+
+func (b *base) Closed() bool {
+	b.closedMu.Lock()
+	defer b.closedMu.Unlock()
+	return b.closed
+}
+
+// A sequentialLayerResolver background fetches spans sequentially, starting from span 0.
+type sequentialLayerResolver struct {
+	*base
+	nextSpanFetchID soci.SpanID
+}
+
+func NewSequentialResolver(layerDigest digest.Digest, spanManager *sm.SpanManager) Resolver {
+	return &sequentialLayerResolver{
+		base: &base{
+			SpanManager: spanManager,
+			layerDigest: layerDigest,
+		},
+	}
+}
+
+func (lr *sequentialLayerResolver) Resolve(ctx context.Context) (bool, error) {
+	log.G(ctx).WithTime(time.Now()).WithFields(logrus.Fields{
+		"layer":  lr.layerDigest,
+		"spanId": lr.nextSpanFetchID,
+	}).Debug("fetching span in background")
```

Comment:
I'm inclined to not mention "background" in this. This function itself shouldn't aware whether this is used in background or not.